app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistViewModel.kt

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.privacy.ui
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModel
+import com.duckduckgo.app.global.DefaultDispatcherProvider
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.SingleLiveEvent
+import com.duckduckgo.app.global.UriString
+import com.duckduckgo.app.privacy.db.UserWhitelistDao
+import com.duckduckgo.app.privacy.model.UserWhitelistedDomain
+import com.duckduckgo.app.privacy.ui.WhitelistViewModel.Command.*
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+class WhitelistViewModel(
+    private val dao: UserWhitelistDao,
+    private val dispatchers: DispatcherProvider = DefaultDispatcherProvider()
+) : ViewModel() {
+
+    data class ViewState(
+        val showWhitelist: Boolean = true,
+        val whitelist: List<UserWhitelistedDomain> = emptyList()
+    )
+
+    sealed class Command {
+        object ShowAdd : Command()
+        class ShowEdit(val entry: UserWhitelistedDomain) : Command()
+        class ConfirmDelete(val entry: UserWhitelistedDomain) : Command()
+        object ShowWhitelistFormatError : Command()
+    }
+
+    val viewState: MutableLiveData<ViewState> = MutableLiveData()
+    val command: SingleLiveEvent<Command> = SingleLiveEvent()
+
+    private val entries: LiveData<List<UserWhitelistedDomain>> = dao.all()
+    private val observer = Observer<List<UserWhitelistedDomain>> { onUserWhitelistChanged(it!!) }
+
+    init {
+        viewState.value = ViewState()
+        entries.observeForever(observer)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        entries.removeObserver(observer)
+    }
+
+    private fun onUserWhitelistChanged(entries: List<UserWhitelistedDomain>) {
+        viewState.value = viewState.value?.copy(
+            showWhitelist = entries.isNotEmpty(),
+            whitelist = entries
+        )
+    }
+
+    fun onAddRequested() {
+        command.value = ShowAdd
+    }
+
+    fun onEntryAdded(entry: UserWhitelistedDomain) {
+        if (!UriString.isValidDomain(entry.domain)) {
+            command.value = ShowWhitelistFormatError
+            return
+        }
+        GlobalScope.launch(dispatchers.io()) {
```

Comment:
_I don't have the full context of the code which calls this function, so bear that in mind when considering my answer here._

Generally, the problem with `GlobalScope.launch` here, or any launching from any scope at all, is the risk of creating a race condition, whereby `onEntryAdded()` has finished executing but the launched coroutine is still executing. The risk is that we base logic on `onEntryAdded()` finishing and assume the entry has in fact been added, when in reality it might have been added already, might still to be added or be in the middle of being added; we have no way of knowing from the calling code.

if `onEntryAdded` were a `suspend` function, we'd have more control over this but I don't know how that would affect other code and if we need that control. we've been burnt a few times because of similar problems, so it would certainly be good to start tightening these things up instead of introducing more of them. But again, i'm not familiar with the larger changes introduced here. Happy to dig in more if required, or discuss offline if folks wanted to sync up about it.

Reply:
Thanks @CDRussell, that's a very nice explanation for the things to take into consideration when using `launch`.

I was not proposing to replace `launch` but to change the `coroutineScope`. Is there any recommendation or things to validate before using `globalScope` instead of `viewmodelScope`?

Reply:
We discussed this over zoom and concluded that:
* keeping GlobalScope is ok as we want the operation to live on past the 
* the db methods should be suspend functions for clarity and to avoid people using them without considering async nature as this method grows - this has been updated